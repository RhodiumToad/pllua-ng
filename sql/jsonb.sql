--

\set VERBOSITY terse

--

do language pllua $$

  a = { json_test = "This is only a test.",
        foo = "If this were real data, it would make more sense.",
	piem = [[
Now I, even I, would celebrate
In rhymes unapt the great
Immortal Syracusan rivaled nevermore,
Who in his wondrous lore,
Passed on before,
Left men his guidance
How to circles mensurate.
]],
        names = { "Dougal", "Florence", "Ermintrude", "Zebedee",
	          "Brian", "Dylan" },
	mixed = { nil, nil, 123, "foo", nil, true, false, [23] = "fred" },
	empty = {},
	empty2 = {{},{}},
	nested = { "arrayelem", { ["object key"] = "object val",
	                          subobject = { subarray = { { { 123 } } } } } }
      }
  b = pgtype.jsonb(a)
  print(b)
  c = pgtype.jsonb(a, { null = false })
  print(c)
  d = pgtype.jsonb(a, { map = function(v) if type(v) == "boolean" then v = tostring(v) end return v end })
  print(d)
  e = pgtype.jsonb(a, { array_thresh = 1 })
  print(e)
  f = pgtype.jsonb(a, { empty_object = true })
  print(f)

  for k,v in pairs(b) do
    print(k,type(v),v)
    if k == "mixed" then
      for k2,v2 in pairs(v) do print("",k2,type(v2),v2) end
    end
  end

  b { map = print, norecurse = true, pg_numeric = true, discard = true }

  spi.execute([[ create temp table jt1 as select $1 as a ]], b)

$$;

select a, pg_typeof(a) from jt1;

create temp table jt2(id serial, a jsonb);
insert into jt2(a) values ('1');
insert into jt2(a) values ('"foo"');
insert into jt2(a) values ('true');
insert into jt2(a) values ('null');
insert into jt2(a) values ('{"foo":123}');
insert into jt2(a) values ('{"foo":null}');
insert into jt2(a) values ('[10,20,30]');
insert into jt2(a) values ('{"foo":[2,4,6]}');
insert into jt2(a) values ('[{"foo":"bar"},{"baz":"foo"},123,null]');
-- check objects with keys that look like numbers
insert into jt2(a) values ('{"1":"foo", "2":[false,true], "foo":{}}');
insert into jt2(a) values ('{"1":"foo", "2":[false,true]}');

do language pllua $$
  s = spi.prepare([[ select a from jt2 order by id ]])
  for r in s:rows() do
    print(r.a)
    b = r.a(function(k,v,...)
              if type(v)~="table" then
	        print("mapfunc",type(k),k,v,...)
	      else
	        print("mapfunc",type(k),k,type(v),...)
	      end
	      return k,v
	    end)
    print(type(b))
  end
$$;

create temp table jt3(id integer, a jsonb);
-- first row should be plain, then a couple with compressed values,
-- then a couple with external toast
insert into jt3 select i, ('[' || repeat('"foo",',10*(10^i)::integer) || i || ']')::jsonb from generate_series(1,5) i;
do language pllua $$
  s = spi.prepare([[ select a from jt3 where id = $1 ]])
  for i = 1,5 do
    local r = (s:execute(i))[1]
    local a = r.a()
    print(#a,a[#a])
  end
$$;

-- test jsonb in jsonb and similar paths

do language pllua $$
  local jtst1 = pgtype.jsonb('"foo"')   -- json scalar
  local jtst2 = pgtype.jsonb('{"foo":true,"bar":[1,2,false]}')   -- json container
  local ts1 = pgtype.timestamp('2017-12-19 12:00:00')
  print(pgtype.jsonb({ v1 = jtst1,
                       v2 = jtst2,
		       v3 = ts1 }))
$$;

-- test round-trip conversions

do language pllua $$
  local j_in = pgtype.jsonb('{"foo":[1,null,false,{"a":null,"b":[]},{},[]]}')
  local nvl = {}
  local val = j_in{ null = nvl }
  local j_out = pgtype.jsonb(val, { null = nvl })
  print(j_in)
  print(j_out)
$$;

--end
