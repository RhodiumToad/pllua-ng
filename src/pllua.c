/*
 * pllua.c: PL/Lua NG call handler
 * By Andrew "RhodiumToad" Gierth, rhodiumtoad at postgresql.org
 * Based in some part on pllua by Luis Carvalho and others
 * License: MIT license or PostgreSQL licence
 */

#include "pllua.h"

#include "commands/event_trigger.h"
#include "commands/trigger.h"

PG_MODULE_MAGIC;

/*
 * Exposed interface
 *
 * see also _PG_init in init.c
 */
PGDLLEXPORT Datum pllua_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_inline_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_inline_handler(PG_FUNCTION_ARGS);

static Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted);


/* Trusted entry points */

PG_FUNCTION_INFO_V1(pllua_validator);
Datum pllua_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_call_handler);
Datum pllua_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_inline_handler);
Datum pllua_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, true);
}

/* Untrusted entry points */

PG_FUNCTION_INFO_V1(plluau_validator);
Datum plluau_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_call_handler);
Datum plluau_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_inline_handler);
Datum plluau_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, false);
}

/* Common implementations */

static void pllua_entry_stack_check(void)
{
	check_stack_depth();
}

Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	pllua_func_activation *funcact = (fcinfo->flinfo) ? fcinfo->flinfo->fn_extra : NULL;
	ErrorContextCallback ecxt;

	pllua_entry_stack_check();

	act.fcinfo = fcinfo;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = InvalidOid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = NULL;

#if PG_VERSION_NUM >= 110000
	if (fcinfo->context && IsA(fcinfo->context, CallContext))
		act.atomic = castNode(CallContext, fcinfo->context)->atomic;
#endif

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		if (funcact && funcact->thread)
			act.interp = funcact->interp;
		else
			act.interp = pllua_getstate(trusted, &act);

		interp = act.interp;

		if (funcact && funcact->thread)
		{
			/*
			 * We're resuming a value-per-call SRF, so we bypass almost
			 * everything since we don't want to, for example, compile a new
			 * version of the function halfway through a result set. We know
			 * we're in a non-first-row condition if there's an existing thread
			 * in the function activation.
			 */
			pllua_initial_protected_call(act.interp, pllua_resume_function, &act);
		}
		else if (CALLED_AS_TRIGGER(fcinfo))
			pllua_initial_protected_call(act.interp, pllua_call_trigger, &act);
		else if (CALLED_AS_EVENT_TRIGGER(fcinfo))
			pllua_initial_protected_call(act.interp, pllua_call_event_trigger, &act);
		else
			pllua_initial_protected_call(act.interp, pllua_call_function, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	return act.retval;
}

Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	Oid funcoid = PG_GETARG_OID(0);
	ErrorContextCallback ecxt;

	/* security checks */
	if (!CheckFunctionValidatorAccess(fcinfo->flinfo->fn_oid, funcoid))
		PG_RETURN_VOID();

	/*
	 * This doesn't need a stack check because the validator is careful
	 * not to execute user-supplied Lua code.
	 */

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = funcoid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = NULL;

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		interp = act.interp = pllua_getstate(trusted, &act);

		pllua_initial_protected_call(act.interp, pllua_validate, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	PG_RETURN_VOID();
}

Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	ErrorContextCallback ecxt;

	pllua_entry_stack_check();

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = (InlineCodeBlock *) PG_GETARG_POINTER(0);
	act.validate_func = InvalidOid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = "inline block entry";

#if PG_VERSION_NUM >= 110000
	act.atomic = act.cblock->atomic;
#endif

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/* probably excess paranoia */
	if (act.cblock->langIsTrusted != act.trusted)
		elog(ERROR, "trusted state mismatch");

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		interp = act.interp = pllua_getstate(trusted, &act);

		pllua_initial_protected_call(act.interp, pllua_call_inline, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	PG_RETURN_VOID();
}
