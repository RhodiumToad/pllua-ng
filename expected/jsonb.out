--
\set VERBOSITY terse
--
do language pllua $$

  a = { json_test = "This is only a test.",
        foo = "If this were real data, it would make more sense.",
	piem = [[
Now I, even I, would celebrate
In rhymes unapt the great
Immortal Syracusan rivaled nevermore,
Who in his wondrous lore,
Passed on before,
Left men his guidance
How to circles mensurate.
]],
        names = { "Dougal", "Florence", "Ermintrude", "Zebedee",
	          "Brian", "Dylan" },
	mixed = { nil, nil, 123, "foo", nil, true, false, [23] = "fred" },
	empty = {},
	empty2 = {{},{}},
	nested = { "arrayelem", { ["object key"] = "object val",
	                          subobject = { subarray = { { { 123 } } } } } }
      }
  b = pgtype.jsonb(a)
  print(b)
  c = pgtype.jsonb(a, { null = false })
  print(c)
  d = pgtype.jsonb(a, { map = function(v) if type(v) == "boolean" then v = tostring(v) end return v end })
  print(d)
  e = pgtype.jsonb(a, { array_thresh = 1 })
  print(e)
  f = pgtype.jsonb(a, { empty_object = true })
  print(f)

  spi.execute([[ create temp table jt1 as select $1 as a ]], b)

$$;
INFO:  {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": [], "mixed": [null, null, 123, "foo", null, true, false, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, "fred"], "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [[], []], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."}
INFO:  {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": [], "mixed": [null, null, 123, "foo", null, true, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, "fred"], "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [[], []], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."}
INFO:  {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": [], "mixed": [null, null, 123, "foo", null, "true", "false", null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, "fred"], "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [[], []], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."}
INFO:  {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": [], "mixed": {"3": 123, "4": "foo", "6": true, "7": false, "23": "fred"}, "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [[], []], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."}
INFO:  {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": {}, "mixed": [null, null, 123, "foo", null, true, false, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, "fred"], "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [{}, {}], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."}
select a, pg_typeof(a) from jt1;
                                                                                                                                                                                                                                                                                                                                         a                                                                                                                                                                                                                                                                                                                                         | pg_typeof 
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------
 {"foo": "If this were real data, it would make more sense.", "piem": "Now I, even I, would celebrate\nIn rhymes unapt the great\nImmortal Syracusan rivaled nevermore,\nWho in his wondrous lore,\nPassed on before,\nLeft men his guidance\nHow to circles mensurate.\n", "empty": [], "mixed": [null, null, 123, "foo", null, true, false, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, "fred"], "names": ["Dougal", "Florence", "Ermintrude", "Zebedee", "Brian", "Dylan"], "empty2": [[], []], "nested": ["arrayelem", {"subobject": {"subarray": [[[123]]]}, "object key": "object val"}], "json_test": "This is only a test."} | jsonb
(1 row)

create temp table jt2(id serial, a jsonb);
insert into jt2(a) values ('1');
insert into jt2(a) values ('"foo"');
insert into jt2(a) values ('true');
insert into jt2(a) values ('null');
insert into jt2(a) values ('{"foo":123}');
insert into jt2(a) values ('{"foo":null}');
insert into jt2(a) values ('[10,20,30]');
insert into jt2(a) values ('{"foo":[2,4,6]}');
insert into jt2(a) values ('[{"foo":"bar"},{"baz":"foo"},123,null]');
-- check objects with keys that look like numbers
insert into jt2(a) values ('{"1":"foo", "2":[false,true], "foo":{}}');
insert into jt2(a) values ('{"1":"foo", "2":[false,true]}');
do language pllua $$
  s = spi.prepare([[ select a from jt2 order by id ]])
  for r in s:rows() do
    print(r.a)
    b = r.a(function(k,v,...)
              if type(v)~="table" then
	        print("mapfunc",type(k),k,v,...)
	      else
	        print("mapfunc",type(k),k,type(v),...)
	      end
	      return k,v
	    end)
    print(type(b))
  end
$$;
INFO:  1
INFO:  mapfunc	nil	nil	1
INFO:  number
INFO:  "foo"
INFO:  mapfunc	nil	nil	foo
INFO:  string
INFO:  true
INFO:  mapfunc	nil	nil	true
INFO:  boolean
INFO:  null
INFO:  mapfunc	nil	nil	nil
INFO:  nil
INFO:  {"foo": 123}
INFO:  mapfunc	string	foo	123
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  {"foo": null}
INFO:  mapfunc	string	foo	nil
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  [10, 20, 30]
INFO:  mapfunc	number	0	10
INFO:  mapfunc	number	1	20
INFO:  mapfunc	number	2	30
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  {"foo": [2, 4, 6]}
INFO:  mapfunc	number	0	2	foo
INFO:  mapfunc	number	1	4	foo
INFO:  mapfunc	number	2	6	foo
INFO:  mapfunc	string	foo	table
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  [{"foo": "bar"}, {"baz": "foo"}, 123, null]
INFO:  mapfunc	string	foo	bar	0
INFO:  mapfunc	number	0	table
INFO:  mapfunc	string	baz	foo	1
INFO:  mapfunc	number	1	table
INFO:  mapfunc	number	2	123
INFO:  mapfunc	number	3	nil
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  {"1": "foo", "2": [false, true], "foo": {}}
INFO:  mapfunc	string	1	foo
INFO:  mapfunc	number	0	false	2
INFO:  mapfunc	number	1	true	2
INFO:  mapfunc	string	2	table
INFO:  mapfunc	string	foo	table
INFO:  mapfunc	nil	nil	table
INFO:  table
INFO:  {"1": "foo", "2": [false, true]}
INFO:  mapfunc	string	1	foo
INFO:  mapfunc	number	0	false	2
INFO:  mapfunc	number	1	true	2
INFO:  mapfunc	string	2	table
INFO:  mapfunc	nil	nil	table
INFO:  table
create temp table jt3(id integer, a jsonb);
-- first row should be plain, then a couple with compressed values,
-- then a couple with external toast
insert into jt3 select i, ('[' || repeat('"foo",',10*(10^i)::integer) || i || ']')::jsonb from generate_series(1,5) i;
do language pllua $$
  s = spi.prepare([[ select a from jt3 where id = $1 ]])
  for i = 1,5 do
    local r = (s:execute(i))[1]
    local a = r.a()
    print(#a,a[#a])
  end
$$;
INFO:  101	1
INFO:  1001	2
INFO:  10001	3
INFO:  100001	4
INFO:  1000001	5
-- test jsonb in jsonb and similar paths
do language pllua $$
  local jtst1 = pgtype.jsonb('"foo"')   -- json scalar
  local jtst2 = pgtype.jsonb('{"foo":true,"bar":[1,2,false]}')   -- json container
  local ts1 = pgtype.timestamp('2017-12-19 12:00:00')
  print(pgtype.jsonb({ v1 = jtst1,
                       v2 = jtst2,
		       v3 = ts1 }))
$$;
INFO:  {"v1": "foo", "v2": {"bar": [1, 2, false], "foo": true}, "v3": "2017-12-19T12:00:00"}
-- test round-trip conversions
do language pllua $$
  local j_in = pgtype.jsonb('{"foo":[1,null,false,{"a":null,"b":[]},{},[]]}')
  local nvl = {}
  local val = j_in{ null = nvl }
  local j_out = pgtype.jsonb(val, { null = nvl })
  print(j_in)
  print(j_out)
$$;
INFO:  {"foo": [1, null, false, {"a": null, "b": []}, {}, []]}
INFO:  {"foo": [1, null, false, {"a": null, "b": []}, {}, []]}
--end
